### PR TITLE
add brief design doc for alps_submitter

### PR DESCRIPTION
**Pull request by uuleon**

### CODE REVIEW

File: doc/alps_submitter.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# ALPS Submitter
+
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md:6]: Could you add a few lines explaining the context of this design doc? For example, what is ALPS? Why sqlflow needs a alps submitter (e.g., ALPS supports distributed training; ALPS has rich data preprocessing functionalities)?

File: doc/alps_submitter.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# ALPS Submitter
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: This is an experimental design. How about changing the title to "Proof of Concept: ALPS Submitter"

File: doc/alps_submitter.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+## Precondition
+1. Custom estimator is not supported.
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: Please change passive voice to active voice such as "For machine learning models, we only consider [Tensorflow premade estimator](https://www.tensorflow.org/guide/premade_estimators)."

File: doc/alps_submitter.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+1. Train with evaluate is not supported.
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: Please change passive voice to active voice.

File: doc/alps_submitter.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+1. Inputs needs to set the type of the dense/sparse, shape, etc.
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: "needs" => "need".

The reader may not know when to use `dense`. How about changing it to 
"If a table cell is encoded, we assume the user always provides enough decoding information such as dense/sparse, shape via expression such as `DENSE`, `SPARSE`, `IDENTITY_SPARSE`."

File: doc/alps_submitter.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+```python
+from alps.feature import FeatureColumnsBuilder
+
+class CustomFCBuilder(FeatureColumnsBuilder):
+  def build_feature_columns():
+    /* code generated here*/
+    return feature_columns_dict
+```
+
+### Estimator Args
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: What is "Estimator Args"?

File: doc/alps_submitter.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+WITH
+  ...
+COLUMN
+    CROSS([NUMERIC(DENSE(c1, shape=[5], dtype=float)), BUCKETIZED(NUMERIC(DENSE(c1, shape=[5], dtype=float)), [0, 10, 100])])
+LABEL class
+...
+```
+
+
+The `as` syntax is temporarily not supported in `COLUMN`, so the case where Estimator requires multiple group of feature columns is not supported at this time.
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: The following could be an excellent feature request in the future. Maybe file it as a separate PR.

File: doc/alps_submitter.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+TRAIN DNNClassifier
+WITH
+    ...
+COLUMN
+    NUMERIC(f1 * 10)
+```
+
+### Feature Columns Code Generation
+```python
+from alps.feature import FeatureColumnsBuilder
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md:178]: I am not sure where is this code snippet used? Should sqlflow generated code include this code snippet?

File: doc/alps_submitter.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+## Precondition
+1. Custom estimator is not supported.
+1. Train with evaluate is not supported.
+1. Inputs needs to set the type of the dense/sparse, shape, etc.
+
+## Data Pipeline        
+Standard Select -> Train Input Table -> Parser -> Input Fn and TrainSpec
+
+Parser is implemented in `COLUMN`, using expr like `DENSE`, `SPARSE`, `IDENTITY_SPARSE`.
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: "implemented" or "declared"? Change to "In `COLUMN` clause we declare parsing type such as ... "?

File: doc/alps_submitter.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+select 
+  c1, c2, c3, c4, c5 as class
+from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+  batch_size = 512
+  train_spec.max_step = 100
+  ...
+COLUMN
+  DENSE(c1, shape=[5], dtype=float)
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: Should `DENSE(c1, shape=[5], dtype=float)` include `separator=","`?

If not, does it mean all dense features are encoded in a string and separated by commas?

File: doc/alps_submitter.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+select 
+    c1, c2, c3, c4, c5 as class
+from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+  batch_size = 512
+  train_spec.max_step = 100
+  ...
+COLUMN
+  CROSS([NUMERIC(DENSE(c1, shape=[5], dtype=float)), BUCKETIZED(NUMERIC(DENSE(c1, shape=[5], dtype=float)), [0, 10, 100])])
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md]: How are `CROSS`, `NUMERIC` and `BUCKETIZED` translated into `graph.conf`?

File: doc/alps_submitter.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+LABEL class
+...
+```
+
+### Semantic Analyze
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md:160]: I am not sure how can we implement this feature. Do you have a concrete design?

File: doc/alps_submitter.md (hunk 11 above)
Comment (uuleon) [reply doc/alps_submitter.md]: We don't need to translate these expressions to `graph.conf`,  they should be translated to a snippet code using [tensorflow feature columns api](https://www.tensorflow.org/guide/feature_columns). ALPS framework will execute it and pass the result to the constructor method of the estimator.

File: doc/alps_submitter.md (hunk 12 above)
Comment (uuleon) [reply doc/alps_submitter.md:160]: Sorry, we have no concrete design right now. This part may be implemented by an AST checker or something else. After the simplest case is done, we will give a supplement design.

File: doc/alps_submitter.md (hunk 8 above)
Comment (uuleon) [reply doc/alps_submitter.md:178]: ALPS framework will execute this code snippet and pass the result to the constructor method of the estimator.

File: doc/alps_submitter.md (hunk 6 above)
Comment (uuleon) [reply doc/alps_submitter.md]: Sorry for the mistake, it's actually the parameters of the constructor method of the estimator.

File: doc/alps_submitter.md (hunk 7 above)
Comment (uuleon) [reply doc/alps_submitter.md]: Yes. So it is not supported in this design.

**Comment by tonyyang-svail** [comment]
To make CI pass, please merge/rebase develop branch.

File: doc/alps_submitter.md (hunk 1 above)
Comment (uuleon) [reply doc/alps_submitter.md:6]: done

File: doc/alps_submitter.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+<b>SQL</b>
+```sql
+select 
+  c1, c2, c3 as class
+from kaggle_credit_fraud_training_data
+TRAIN DNNClassifier
+WITH
+  ...
+COLUMN
+  DENSE(c1, shape=3, dtype=float, delimiter=',')
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md:96]: `DENSE(c1, shape=3, dtype=float, delimiter=',')`
or
`DENSE(c1, 3, float, ',')`

File: doc/alps_submitter.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+WITH
+    estimator.hidden_units = [10, 20],
+    train_spec.max_steps = 2000,
+    input_fn.batch_size = 512
+COLUMN
+    CROSS([NUMERIC(c1), BUCKETIZED(NUMERIC(c2), [0, 10, 100])])
+LABEL class
+...
+```
+
```
Comment (tonyyang-svail) [inline doc/alps_submitter.md:219]: How about changing the following sentence to

## Future Development

For now, we will combine the `feature_columns` as one dense tensor and feed it to the model. If we want to support models with multiple inputs, we will need an `AS` keyword annotate different inputs.
